### PR TITLE
Improve advanced reboot test to support multiple VLANs

### PR DESCRIPTION
**Pull request by shi-su**
### Description of PR

Summary: Improve advanced reboot test to support multiple VLANs and the scenario with portchannel in VLAN
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911

#### What is the motivation for this PR?
The current advanced reboot tests only support the scenario with one vlan and no portchannel in vlans. This PR aims to improve the test cases to cover broader scenarios.

#### How did you do it?
Update advanced reboot to take account of multiple VLANs and ignore portchannel ports in VLANs.

### CODE REVIEW

File: ansible/roles/test/files/ptftests/advanced-reboot.py [hunk 1]
```diff
@@ -137,7 +138,6 @@ def __init__(self):
         self.check_param('vlan_ports_file', '', required=True)
         self.check_param('ports_file', '', required=True)
         self.check_param('dut_mac', '', required=True)
-        self.check_param('dut_vlan_ip', '', required=True)
```
Comment (qiluo-msft) [inline ansible/roles/test/files/ptftests/advanced-reboot.py:140]: >check_param [](http://example.com/codeflow?start=13&length=11)

Could you still check_param?

File: ansible/roles/test/files/ptftests/advanced-reboot.py [hunk 2]
```diff
@@ -317,19 +327,21 @@ def timeout(self, func, seconds, message):
 
     def generate_vlan_servers(self):
         vlan_host_map = defaultdict(dict)
-        vlan_ip_range = self.test_params['vlan_ip_range']
-
-        _, mask = vlan_ip_range.split('/')
-        n_hosts = min(2**(32 - int(mask)) - 3, self.max_nr_vl_pkts)
+        self.nr_vl_pkts = 0
```
Comment (qiluo-msft) [inline ansible/roles/test/files/ptftests/advanced-reboot.py:330]: >nr_vl_pkts [](http://example.com/codeflow?start=13&length=10)

The var name is hard to read. Could you add some code comment?

File: ansible/roles/test/files/ptftests/advanced-reboot.py [hunk 3]
```diff
@@ -317,19 +327,21 @@ def timeout(self, func, seconds, message):
...
-            mac = self.VLAN_BASE_MAC_PATTERN.format(counter)
-            port = self.vlan_ports[i % len(self.vlan_ports)]
-            addr = self.host_ip(vlan_ip_range, i)
+            for counter, i in enumerate(xrange(2, n_hosts + 2)):
+                mac = self.VLAN_BASE_MAC_PATTERN.format(counter)
+                port = self.ports_per_vlan[vlan][i % len(self.ports_per_vlan[vlan])]
+                addr = self.host_ip(prefix, i)
 
-            vlan_host_map[port][addr] = mac
+                vlan_host_map[port][addr] = mac     # For simplicity, assuming there is no overlapped ports between vlans
```
Comment (qiluo-msft) [inline ansible/roles/test/files/ptftests/advanced-reboot.py:342]: >assuming [](http://example.com/codeflow?start=70&length=8)

Why assume this?

File: ansible/roles/test/files/ptftests/advanced-reboot.py (hunk 2 above)
Comment (shi-su) [reply ansible/roles/test/files/ptftests/advanced-reboot.py]: Added comments to clarify this is the number of packets from upper layer.

File: ansible/roles/test/files/ptftests/advanced-reboot.py (hunk 3 above)
Comment (shi-su) [reply ansible/roles/test/files/ptftests/advanced-reboot.py]: I checked this part again, it seems that it does not necessarily require the assumption. Removed the comment.

File: ansible/roles/test/files/ptftests/advanced-reboot.py [hunk 4]
```diff
@@ -394,10 +406,17 @@ def build_peer_mapping(self):
...
-        if len(content) > 1:
-            raise Exception("Too many vlans")
-        return [(ifname, self.port_indices[ifname]) for ifname in content.values()[0]['members']]
+        portchannel_content = self.read_json('portchannel_ports_file')
+        portchannel_names = [pc['name'] for pc in portchannel_content.values()]
+
+        vlan_content = self.read_json('vlan_ports_file')
+        
+        vlan_if_port = []
+        for vlan in self.vlan_ip_range.keys():
```
Comment (qiluo-msft) [inline ansible/roles/test/files/ptftests/advanced-reboot.py:415]:
```suggestion
    for vlan in self.vlan_ip_range:
```
Not a blocking issue, but you can make your code shorter and more pythonic.

Comment (shi-su) [reply ansible/roles/test/files/ptftests/advanced-reboot.py]: Thanks for the comment, fixed as suggested.

**Review by qiluo-msft — APPROVED** [review]

**Pull request merged**